pkg/streamingpromql/operators/binary_operation.go

```diff
@@ -512,45 +562,184 @@ func (b *BinaryOperation) mergeOneSide(data []types.InstantVectorSeriesData, sou
+		sourceSeriesIndexInData := -1
+
+		for seriesIndexInData, d := range data {
+			if len(d.Histograms) == 0 {
+				continue
+			}
+
+			nextPointInSeries := d.Histograms[0]
+			if nextPointInSeries.T == nextT {
+				// Another series has a point with the same timestamp. We have a conflict.
+				firstConflictingSeriesLabels := sourceSeriesMetadata[sourceSeriesIndices[sourceSeriesIndexInData]].Labels
+				secondConflictingSeriesLabels := sourceSeriesMetadata[sourceSeriesIndices[seriesIndexInData]].Labels
+				groupLabels := b.labelsFunc()(firstConflictingSeriesLabels)
 
-	// For one-to-one matching for arithmetic operators, reuse one of the input slices to avoid allocating another slice.
-	// We'll never produce more points than the smaller input side, so use that as our output slice.
+				return nil, fmt.Errorf("found duplicate series for the match group %s on the %s side of the operation at timestamp %s: %s and %s", groupLabels, side, timestamp.Time(nextT).Format(time.RFC3339Nano), firstConflictingSeriesLabels, secondConflictingSeriesLabels)
+			}
+
+			if d.Histograms[0].T < nextT {
+				nextT = d.Histograms[0].T
+				sourceSeriesIndexInData = seriesIndexInData
+			}
+		}
+
+		histogramOutput = append(histogramOutput, data[sourceSeriesIndexInData].Histograms[0])
+		data[sourceSeriesIndexInData].Histograms = data[sourceSeriesIndexInData].Histograms[1:]
+
+		if len(data[sourceSeriesIndexInData].Histograms) == 0 {
+			remainingSeries--
+		}
+	}
+}
+
+func (b *BinaryOperation) computeResult(left types.InstantVectorSeriesData, right types.InstantVectorSeriesData) (types.InstantVectorSeriesData, error) {
+	var fPoints []promql.FPoint
+	var hPoints []promql.HPoint
+
+	// For one-to-one matching for arithmetic operators, We'll never produce more points than the smaller input side.
+	// Because histograms and floats can be multiplied together, we use the sum of both the histogram and float points.
+	// We also don't know if the output will be exclusively floats or histograms, so we'll use the same size slice for both.
+	// We only assign the slices once we see the associated point type so it shouldn't be common that we allocate both.
 	//
 	// FIXME: this is not safe to do for one-to-many, many-to-one or many-to-many matching, as we may need the input series for later output series.
-	if len(left.Floats) < len(right.Floats) {
-		output = left.Floats[:0]
-		defer b.Pool.PutFPointSlice(right.Floats)
+	var leftPutFPointSlice, leftPutHPointSlice, rightPutFPointSlice, rightPutHPointSlice bool
+	leftPoints := len(left.Floats) + len(left.Histograms)
+	rightPoints := len(right.Floats) + len(right.Histograms)
+	leftSmaller := leftPoints < rightPoints
+	var maxPoints int
+	if leftSmaller {
+		maxPoints = leftPoints
 	} else {
-		output = right.Floats[:0]
-		defer b.Pool.PutFPointSlice(left.Floats)
+		maxPoints = rightPoints
+	}
+
+	getFSlice := func() error {
+		if maxPoints <= len(left.Floats) && (leftSmaller || maxPoints > len(right.Floats)) {
```

Comment:
`leftSmaller` isn't quite right here: we just care about the size of `left.Floats` relative to `right.Floats`, given those are the slices we'll reuse - it doesn't matter how large the `HPoint` slices are.

It might also be worthwhile introducing some variables to make the logic clearer.

(same applies to `getHSlice` below)


```suggestion
		canFitOutputInLeftSlice := maxPoints <= len(left.Floats)
		canFitOutputInRightSlice := maxPoints <= len(right.Floats)
		
		if canFitOutputInLeftSlice && len(left.Floats) < len(right.Floats) {
			...
		} else if canFitOutputInRightSlice {
			...
		} else {
			...
		}
```

Reply:
I was optimising for a case which I now realise doesn't exist. It could happen if two Histograms outputted a float, but none of the operations do that.